### ISSUE TITLE
Matrix Vector multiply not parallelized.

### ISSUE THREAD

**Issue by bowu** [tensorflow/tensorflow#6752]
### What is the problem?

Matrix vector multiply is not parallelized. Please see my example code. No matter how I change the intra_op_parallelism_threads, the running time is always similar. I used "top" to confirm that only one thread was used. However, the parallel speedups for square matrix matrix multiply are quite noticeable. Again, "top" confirmed that multiple threads were used.

### Environment info
Operating System: Ubuntu 16.04 

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

If installed from binary pip package, provide:

1. A link to the pip package you installed:
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)
2e22f1b20fdfa77b1332c518617391dc32359c5b
2. The output of `bazel version`
0.4.3

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

import tensorflow as tf
import numpy as np
import time

n = 10000

#approach 1:

matrix1 = tf.constant(np.ones(n*n), shape = [n,n])
matrix2 = tf.constant(np.ones(n*1), shape = [n,1])

product1 = tf.matmul(matrix1, matrix2)

start = time.time()
sess = tf.Session(config=tf.ConfigProto(
    inter_op_parallelism_threads=1,
    intra_op_parallelism_threads=12))
sess.run(product1)
end = time.time()
print('\n Approach 1 took: %s%%' % (end - start))

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

**Comment by yaroslavvb** [comment]
Here's a better benchmark that removes the cost of data transfer. It seems CPU usage is indeed quite low and it won't break 6 G ops/sec on Xeon. However, this problem is bandwidth limited so you are going to spend most of the time transferring data between RAM and CPU, so I'm not sure if this performance can be improved much. @benoitsteiner 


```
import os
import sys
#os.environ["CUDA_VISIBLE_DEVICES"]="3"
import tensorflow as tf
import time

n = 8*8192
with tf.device("/cpu:0"):
    matrix1 = tf.ones((n, n))
    matrix2 = tf.ones((n, 1))
    product1 = tf.matmul(matrix1, matrix2)

graph_options=tf.GraphOptions(optimizer_options=tf.OptimizerOptions(opt_level=tf.OptimizerOptions.L0))
sess = tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=int(sys.argv[1]), graph_options=graph_options))

# pre-warming
sess.run(product1.op)

start = time.time()
sess.run(product1.op)
end = time.time()
ops = n**2 + n*(n-1) # n*(n-1) additions, n^2 multiplications
elapsed = (end - start)
rate = ops/elapsed/10**9
print('\n Approach 1 took: %.2f sec, %.2f G ops/sec' % (elapsed, rate,))

```

**Comment by bowu** [comment]
Did you also observe that the operation is not parallelized by Eigen? Please note that this is a different question from whether the operation has high throughput.

**Comment by yaroslavvb** [comment]
I can't tell if it's parallelized or not. I think the goal should be performance, if higher performance is not possible with parallelization, then it shouldn't be parallelized.

**Comment by yaroslavvb** [comment]
What kind of performance are expecting/needing for this operation?

**Comment by bowu** [comment]
I agree that performance should be the goal, but I actually want to get a confirmation about whether TF parallelizes matrix vector multiply.

**Comment by yaroslavvb** [comment]
Not sure who handles CPU Eigen implementations, assigned to martin to reassign

**Comment by yaroslavvb** [comment]
OK, there do seem to be some missing optimization opportunities here.

First of all, matrix*vector matmul doesn't seem to be parallelized. Doing nxn by nx1 multiply uses one CPU core and runs at about 5 G ops/sec.

For some comparative experiments, note that this operation can be equivalently written as tf.mul+tf.reduce_sum, or simply as reduce_sum(...,axis=1) since our vector is all ones

Some experiments with 16 threads and n=8*8192

tf.matmul(matrix, vector) -- utilizes 1 core, 5 G ops/sec
tf.reduce_sum(matrix, axis=0) -- utilizes all cores, 5 G ops/sec
tf.reduce_sum(matrix, axis=1) -- utilizes all cores, 26 G ops/sec
tf.reduce_sum(tf.mul(matrix, vector), axis=1) -- utilizes all cores, 0.6 G ops/sec


@learyg -- is this something that XLA optimization would help speed up?

**Comment by learyg** [comment]
Yeah, I believe @meheffernan had some significant matrix-vector speedups in his queue at some point, not sure we found the priority to finish landing them at the time. +@eliben

**Comment by rmlarsen** [comment]
It is correct that the Eigen implementation deliberately falls back to a sequential implementation for matrix-vector multiplication. The reason is that there are specialized sequential matvec kernels in core Eigen that are approximately 4x faster than the matmul kernel called by the parallelized tensor contraction code used for matmul in TensorFlow.

In other words, if using the matmul kernel, you'd typically have to burn >4x CPU cycles to match the performance of the sequential kernels. Based on some real-world use cases we decided to stick with this and rely on model-level parallelism. 

I agree that there is room for improvement and that it would be awesome if XLA could contribute better code for this.

Top-level logic in tensorflow (batch_matmul is the new matmul):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/batch_matmul_op_impl.h#L212

matvec kernels: https://bitbucket.org/eigen/eigen/src/e6d6c87e1e27ba39a8f4484366ead7d31e6e97a4/Eigen/src/Core/products/GeneralMatrixVector.h?at=default&fileviewer=file-view-default#GeneralMatrixVector.h-31

matmul block panel kernels:
https://bitbucket.org/eigen/eigen/src/e6d6c87e1e27ba39a8f4484366ead7d31e6e97a4/Eigen/src/Core/products/GeneralBlockPanelKernel.h?at=default&fileviewer=file-view-default

**Comment by rmlarsen** [comment]
I should also mention that @benoitsteiner  and others have been working on new low level kernels based on libxsmm. I believe the current focus is to improve matmul performance on Haswell and newer processors. But, once the code is in, we should probably re-evaluate the situation for matvec as well.

**Comment by bowu** [comment]
Great. Thanks.

**Comment by interfect** [comment]
@rmlarsen @benoitsteiner Did the new libxsmm kernels ever happen? Can matrix-vector multiplies be made to run in parallel, even at the expense of total throughput?

I'm trying to speed up a CPU workload (GPT2 generation for https://github.com/AIDungeon/AIDungeon/issues/58), and it looks like a bunch of MatMul operations that a GPU would churn through are still running single-threaded on my multi-core CPU. Even if doing the multiplies across multiple cores wasted a lot of CPU, for this application (an interactive text adventure running on a user's desktop machine), it's worth burning CPU and reducing throughput to achieve lower latency.
